src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -1398,5 +1384,131 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
 
             return new ImmutableArray<T>(tmp);
         }
+
+        private ICollection<int>? FindOrderedIndicesToRemoveByMultiSet(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
+        {
+            var multiSet = new MultiSet(equalityComparer);
+            foreach (T item in items)
+            {
+                multiSet.Add(item);
+            }
+
+            return GetOrderedIndicesToRemoveFor(multiSet);
+        }
+
+        private ICollection<int>? FindOrderedIndicesToRemoveByMultiSet(ReadOnlySpan<T> items, IEqualityComparer<T>? equalityComparer)
+        {
+            var multiSet = new MultiSet(equalityComparer);
+            foreach (ref readonly T item in items)
+            {
+                multiSet.Add(item);
+            }
+
+            return GetOrderedIndicesToRemoveFor(multiSet);
+        }
+
+        private ICollection<int>? GetOrderedIndicesToRemoveFor(MultiSet multiSet)
+        {
+            List<int>? indicesToRemove = null;
+            for (int i = 0; i < array!.Length; i++)
+            {
+                if (multiSet.TryRemove(array[i]))
+                {
+                    (indicesToRemove ??= new()).Add(i);
+                }
+            }
+
+            return indicesToRemove;
+        }
+
+        private SortedSet<int> FindOrderedIndicesToRemove(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
+        {
+            var indicesToRemove = new SortedSet<int>();
+            foreach (T item in items)
+            {
+                int index = -1;
+                do
+                {
+                    index = IndexOf(item, index + 1, equalityComparer);
+                } while (index >= 0 && !indicesToRemove.Add(index) && index < Length - 1);
+            }
+
+            return indicesToRemove;
+        }
+
+        private readonly struct MultiSet
+        {
+            private readonly Dictionary<NullableKeyWrapper, int> _dictionary;
+
+            public MultiSet(IEqualityComparer<T>? equalityComparer)
+            {
+                _dictionary = new Dictionary<NullableKeyWrapper, int>(new NullableKeyWrapperEqualityComparer(equalityComparer));
+            }
+
+            public void Add(T? item)
+            {
+#if NET6_0_OR_GREATER
+                ref int count = ref CollectionsMarshal.GetValueRefOrAddDefault(_dictionary, item, out _);
+                count++;
```

Comment:
Might need to guard against overflow?
```suggestion
                checked { count++; }
```